gear/legacy_identifier_transfer/src/python/legacy_identifier_transfer_app/main.py

```diff
@@ -0,0 +1,212 @@
+"""Defines legacy_identifier_transfer."""
+
+from datetime import datetime
+import logging
+from typing import Dict, Mapping
+
+from pydantic import ValidationError
+
+from enrollment.enrollment_project import EnrollmentProject
+from enrollment.enrollment_transfer import EnrollmentRecord
+from gear_execution.gear_execution import GearExecutionError
+from identifiers.model import CenterIdentifiers, IdentifierObject
+from outputs.errors import ErrorWriter, ListErrorWriter, identifier_error, legacy_naccid_error, unexpected_value_error
+
+log = logging.getLogger(__name__)
+
+
+class LegacyEnrollmentBatch:
+    """Handles batch processing of legacy enrollment records."""
+
+    def __init__(self) -> None:
+        self.__records: Dict[str, EnrollmentRecord] = {}
+
+    def add(self, record: EnrollmentRecord) -> None:
+        """Adds an enrollment record to the batch.
+
+        Args:
+            record: EnrollmentRecord to add to the batch
+        """
+        if record.naccid:  # We know this will exist for legacy records
+            self.__records[record.naccid] = record
+        else:
+            log.warning('Skipping record with missing NACCID: %s', record)
+
+    def __len__(self) -> int:
+        return len(self.__records)
+
+    def __iter__(self):
+        return iter(self.__records.values())
+
+
+def process_legacy_identifiers(
```

Comment:
My general take is this function is too long and complex.
It might get better if you drop the error writer, but I think it would still be too long.
Maybe factor out some other functions that do task on a single identifier.